src/details/hook_rooting.h

```diff
@@ -0,0 +1,298 @@
+//
+template<typename T>
+struct ReturnTypeTraits {};
+
+template<typename R, typename T, typename... HookArgs>
+struct ReturnTypeTraits<R(T::*)(HookArgs...)> {
+  typedef R Type;
+};
+
+// The _HOOK_TASK macro defines an auxiliary classes HookTask_... that invoke
+// a specific plugin hook method with a provided set of method arguments. The
+// HookTask_... classes are meant to be used by OrderedPlugins's apply function
+// called in the static functions of kaleidoscope::Hooks to forward the call to
+// the hook methods of all registered plugins.
+//
+#define _HOOK_TASK(HOOK_METHOD, CONTINUATION_PREDICATE_TYPE)                   \
+__NN__                                                                         \
+__NN__   struct HookTask_##HOOK_METHOD {                                       \
+__NN__                                                                         \
+__NN__      typedef CONTINUATION_PREDICATE_TYPE ContinuationPredicate;         \
+__NL__                                                                         \
+__NN__      /* Use a traits class to determine the return value type of the    \
+__NN__         hook method.                                                    \
+__NN__       */                                                                \
+__NL__      typedef typename ReturnTypeTraits<                                 \
+__NL__                 decltype(&kaleidoscope::Plugin::HOOK_METHOD)            \
+__NL__              >::Type  ReturnType;                                       \
+__NL__                                                                         \
+__NL__      template<typename Plugin__, typename... Args__>                    \
+__NL__      static ReturnType invoke(Plugin__ &plugin,                         \
+__NL__                                    Args__&&... hook_args)               \
+__NL__      {                                                                  \
+__NN__         /* Assert that hook methods in derived plugins have an          \
+__NN__            call signature.                                              \
+__NN__          */                                                             \
+__NL__         _MATCH_HOOK_TYPE(HOOK_METHOD, Plugin__)                         \
+__NN__                                                                         \
+__NL__         return plugin.HOOK_METHOD(hook_args...);                        \
+__NL__      }                                                                  \
+__NL__   };
+
+// Predicate classes that decides on hook-plugin-loop continuation.
+//
+struct ContinueIfHookReturnsTrue {
+  static bool eval(bool val) {
+    return val;
+  }
+};
+
+struct AlwaysContinue {
+  constexpr static bool eval(bool val) {
+    return true;
+  }
+};
+
+_HOOK_TASK(init, ContinueIfHookReturnsTrue) // Defines HookTask_init
+_HOOK_TASK(eventHandlerHook, AlwaysContinue) // Defines HookTask_eventHandlerHook
```

Comment:
Shouldn't `init` be `AlwaysContinue` (since it is void anyway), and `eventHandlerHook` `ContinueIfHookReturns`?

Reply:
Yes, you are right that's a mistake. Unless ... you want to let plugin initializations block each other, which does not make much sense.